### PR TITLE
Improvements to Env::GetChildren

### PR DESCRIPTION
**Pull request by adamretter**
The main improvement here is to not include `.` or `..` in the results of `Env::GetChildren`. The occurrence of `.` or `..`; it is non-portable, dependent on the Operating System and the File System. See: https://www.gnu.org/software/libc/manual/html_node/Reading_002fClosing-Directory.html

There were lots of duplicate checks spread through the RocksDB codebase previously to skip `.` and `..`. This new removes the need for those at the source.

Also some minor fixes to `Env::GetChildren`:
* Improve error handling in POSIX implementation
* Remove unnecessary array allocation on Windows
* Fix struct name for Windows Non-UTF-8 API

### CODE REVIEW

File: port/win/env_win.cc
Comment (mrambacher) [inline port/win/env_win.cc:595]: Can you change this to 
   ((data.dwFiileAttributes & FILE_ATTRIBUTE) != 0) &&
I think it reads a little clearer.

File: utilities/persistent_cache/persistent_cache_test.cc
Comment (mrambacher) [inline utilities/persistent_cache/persistent_cache_test.cc:53]: Can this entire routine be deleted and replaced with DestroyDr (from file_util.h)?

Comment (adamretter) [reply utilities/persistent_cache/persistent_cache_test.cc:53]: Done.

File: port/win/env_win.cc
Comment (adamretter) [reply port/win/env_win.cc]: Done.

File: env/fs_posix.cc
Comment (adamretter) [reply env/fs_posix.cc]: As you prefer ;-)

File: env/fs_posix.cc
Comment (adamretter) [reply env/fs_posix.cc]: Yes! Fixed.

File: env/env_test.cc
Comment (adamretter) [reply env/env_test.cc]: Done.

File: include/rocksdb/env.h
Comment (pdillinger) [inline include/rocksdb/env.h:285]: I prefer "shall" to "will" when there is an expectation placed on user code (in addition to providing a guarantee).

**Review by pdillinger — COMMENTED** [review]
I like this a lot, though it is a breaking API change. It needs to be logged clearly in HISTORY.md. Failure to comply could cause undefined behavior in RocksDB.

I defer to @anand1976 for approval.

**Comment by pdillinger** [comment]
> I defer to @anand1976 for approval.

There might be some FB-internal follow-up that someone would need to take on.

File: env/env_basic_test.cc
Comment (mrambacher) [inline env/env_basic_test.cc:39]: Can this also be changed into DestroyDir?

File: env/env_basic_test.cc
Comment (mrambacher) [inline env/env_basic_test.cc:332]: Are you guaranteed the test_dir is empty when you enter this method, or do you need to check that the number of children has increased by 1?

**Review by mrambacher — APPROVED** [review]
LGTM!

File: include/rocksdb/env.h
Comment (adamretter) [reply include/rocksdb/env.h:285]: As a native English speaker I was under the impression that the use of `shall` had fallen out of favour; especially within American English. I am certainly no expert in English, but various sources I checked seem to confirm that. Do you still want that change made?

**Comment by adamretter** [comment]
>  It needs to be logged clearly in HISTORY.md

@pdillinger Okay, I have done that now.

File: env/env_basic_test.cc
Comment (adamretter) [reply env/env_basic_test.cc:39]: Done.

File: env/env_basic_test.cc
Comment (adamretter) [reply env/env_basic_test.cc:332]: Yes, I think that `EnvBasicTestWithParam::TearDown` takes care of that.

File: include/rocksdb/env.h
Comment (pdillinger) [reply include/rocksdb/env.h:285]: Wikipedia doesn't lie, right? ;)

https://en.wikipedia.org/wiki/Shall_and_will#Legal_and_technical_use

> Shall and will are distinguished by NASA[18] and Wikiversity[19] as follows:
>
> * Shall is usually used to state a device or system's requirements. For example: "The selected generator shall provide a minimum of 80 Kilowatts."
> * Will is generally used to state a device or system's purpose. For example, "The new generator will be used to power the operations tent."

Comment (anand1976) [reply include/rocksdb/env.h:285]: FWIW, standards documents use "shall" or "shall not" to indicate requirements.

**Review by anand1976 — APPROVED** [review]
LGTM.

**Comment by pdillinger** [comment]
Need merge conflicts resolved

File: include/rocksdb/env.h
Comment (adamretter) [reply include/rocksdb/env.h:285]: @anand1976 I was aware of MUST, COULD, SHOULD, MAY... but not SHALL, e.g. https://tools.ietf.org/html/rfc2119

Comment (adamretter) [reply include/rocksdb/env.h:285]: @pdillinger I have changed it to `shall` as requested.

**Comment by adamretter** [comment]
@pdillinger I rebased and resolved the conflicts etc.

**Comment by pdillinger** [comment]
More conflicts to resolve (not going to do it myself because I assume you have push permission disabled)

**Comment by pdillinger** [comment]
> More conflicts to resolve (not going to do it myself because I assume you have push permission disabled)

Sorry I lost the race to land 😆

**Comment by adamretter** [comment]
Thanks @jay-zhuang and @pdillinger

**Pull request closed without merging**
